analyzer/src/main/kotlin/managers/utils/PackageManagerDependencyHandler.kt

```diff
@@ -0,0 +1,178 @@
+import org.ossreviewtoolkit.model.Package
+import org.ossreviewtoolkit.model.PackageLinkage
+import org.ossreviewtoolkit.model.PackageReference
+import org.ossreviewtoolkit.model.Project
+import org.ossreviewtoolkit.model.jsonMapper
+import org.ossreviewtoolkit.model.utils.DependencyHandler
+import org.ossreviewtoolkit.utils.ort.log
+
+private const val TYPE = "PackageManagerDependency"
+
+class PackageManagerDependencyHandler(
+    private val analyzerResult: AnalyzerResult
+) : DependencyHandler<ResolvableDependencyNode> {
+    companion object {
+        /**
+         * Create a [PackageReference] that points to the result of a [scope] in the [definitionFile] from another
+         * package manager with the provided [linkage]. The analyzer will replace this reference with the dependency
+         * tree of the provided scope after all package managers have finished.
+         */
+        fun createPackageManagerDependency(
+            packageManager: String,
+            definitionFile: String,
+            scope: String,
+            linkage: PackageLinkage
+        ): PackageReference =
+            PackageReference(
+                id = Identifier(
+                    type = TYPE,
+                    namespace = "",
+                    name = jsonMapper.writeValueAsString(
+                        PackageManagerDependency(
+                            packageManager,
+                            definitionFile,
+                            scope,
+                            linkage
+                        )
+                    ).encode(),
+                    version = ""
+                )
+            )
+
+        private fun getPackageManagerDependency(node: DependencyNode): PackageManagerDependency? =
+            when (node.id.type) {
+                TYPE -> jsonMapper.readValue<PackageManagerDependency>(node.id.name.decode())
+                else -> null
+            }
+    }
+
+    private val navigator = DependencyGraphNavigator(analyzerResult.dependencyGraphs)
+
+    override fun createPackage(dependency: ResolvableDependencyNode, issues: MutableList<OrtIssue>): Package? =
+        analyzerResult.packages.find { it.pkg.id == dependency.id }?.pkg
+
+    override fun dependenciesFor(dependency: ResolvableDependencyNode): Collection<ResolvableDependencyNode> =
+        buildList {
+            dependency.visitDependencies { dependencies ->
+                dependencies.forEach { node ->
+                    addAll(resolvePackageManagerDependency(node))
+                }
+            }
+        }
+
+    override fun identifierFor(dependency: ResolvableDependencyNode): Identifier = dependency.id
+
+    override fun issuesForDependency(dependency: ResolvableDependencyNode): Collection<OrtIssue> = dependency.issues
+
+    override fun linkageFor(dependency: ResolvableDependencyNode): PackageLinkage = dependency.linkage
+
+    fun resolvePackageManagerDependency(dependency: DependencyNode): List<ResolvableDependencyNode> =
```

Comment:
This algorithm should be able to deal with nested dependencies to other projects/scopes recursively. How about cycles? Should there be a detection for cyclic dependencies between projects? Is this practically relevant?

Reply:
> > This algorithm should be able to deal with nested dependencies to other projects/scopes recursively.
> 
> I think this should already be the case, as to my understanding the nodes returned by `dependenciesFor` will be handled recursively by the `DependencyGraphBuilder`.

Yes, this is my understanding as well.

> 
> > How about cycles? Should there be a detection for cyclic dependencies between projects? Is this practically relevant?
> 
> Isn't there already cycle prevention built into the `DependencyGraphBuilder`? In general, I currently cannot imagine a situation where such a cycle would be created between different package managers, so maybe we can wait until this actually becomes an issue. I think adding test cases for both scenarios would make sense, but I would prefer to do this in a separate PR.

Not sure whether the cycle detection in the builder is effective here, as `PackageManagerDependencyHandler` also calls itself recursively. But I agree that this is currently a pure theoretically construct and am fine with deferring this to later PRs.